network/src/network.rs

```diff
@@ -189,19 +205,31 @@ impl Network {
         self.local_private_key.to_public_key()
     }
 
-    pub fn external_url(&self) -> Option<String> {
-        self.original_listened_addresses
-            .read()
-            .get(0)
-            .map(|addr| self.to_external_url(addr))
+    pub fn external_urls(&self, max_urls: usize) -> Vec<(String, u8)> {
```

Comment:
We can use `chain` here without allocating two temp Vec.

```suggestion
    pub fn external_urls(&self, max_urls: usize) -> Vec<(String, u8)> {
        let original_listened_addresses = self.original_listened_addresses.read();
        self.listened_addresses(max_urls.saturating_sub(original_listened_addresses.len()))
            .into_iter()
            .chain(
                original_listened_addresses
                    .iter()
                    .map(|addr| (addr.to_owned(), 1)),
            )
            .map(|(addr, score)| (self.to_external_url(&addr), score))
            .collect()
    }
```